otter/test/test_http.py

```diff
@@ -128,8 +146,138 @@ def test_defaults(self):
                     data=None,
                     log=None,
                     reauth_codes=(401, 403),
-                    success_codes=(200,),
+                    success_pred=has_code(200),
                     json_response=True,
                 )
             )
         )
+
+
+class PerformServiceRequestTests(SynchronousTestCase):
+    """Tests for :func:`concretize_service_request`."""
+    def setUp(self):
+        """Save some common parameters."""
+        self.log = object()
+        self.authenticator = object()
+        self.service_mapping = {
+            ServiceType.CLOUD_SERVERS: 'cloudServersOpenStack'}
+        eff = service_request(ServiceType.CLOUD_SERVERS, 'GET', 'servers')
+        self.svcreq = eff.intent
+
+    def _concrete(self, svcreq):
+        """
+        Call :func:`concretize_service_request` with premade test objects.
+        """
+        return concretize_service_request(
+            self.authenticator, self.log, self.service_mapping, 'DFW',
+            1,
+            svcreq
+        )
+
+    def test_authenticates(self):
+        """Auth is done before making the request."""
+        eff = self._concrete(self.svcreq)
+        expected_intent = Authenticate(self.authenticator, 1, self.log)
+        self.assertEqual(eff.intent, expected_intent)
+        next_eff = resolve_authenticate(eff)
+        # The next effect in the chain is the requested HTTP request,
+        # with appropriate auth headers
+        self.assertEqual(
+            next_eff.intent,
+            Request(method='GET', url='http://dfw.openstack/servers',
+                    headers=headers('token'), log=self.log))
+
+    def test_invalidate_on_auth_error_code(self):
+        """
+        Upon authentication error, the auth cache is invalidated.
+        """
+        eff = self._concrete(self.svcreq)
+        next_eff = resolve_authenticate(eff)
+        # When the HTTP response is an auth error, the auth cache is
+        # invalidated, by way of the next effect:
+        invalidate_eff = resolve_effect(next_eff, stub_pure_response("", 401))
+        expected_intent = InvalidateToken(self.authenticator, 1)
+        self.assertEqual(invalidate_eff.intent, expected_intent)
+        self.assertRaises(APIError, resolve_effect, invalidate_eff, None)
+
+    def test_json(self):
+        """
+        JSON-serializable requests are dumped before being sent, and
+        JSON-serialized responses are parsed.
+        """
+        input_json = {"a": 1}
+        output_json = {"b": 2}
+        svcreq = service_request(ServiceType.CLOUD_SERVERS, "GET", "servers",
+                                 data=input_json).intent
+        eff = self._concrete(svcreq)
+        next_eff = resolve_authenticate(eff)
+        result = resolve_effect(next_eff,
+                                stub_pure_response(json.dumps(output_json)))
+        self.assertEqual(next_eff.intent.data, json.dumps(input_json))
```

Comment:
Minor nitpick: It might be easier for reader to move this at line 214 after getting `next_effect`. And a space after the check would be nice. Something like:
```python
# input json is as expected
next_eff = resolve_authenticate(eff)
self.assertEqual(next_eff.intent.data, json.dumps(input_json))

# output json is as expected
result = resolve_effect(next_eff, stub_pure_response(json.dumps(output_json)))
self.assertEqual(result, output_json)
```